hardware/kubernetes_client.go

```diff
@@ -0,0 +1,255 @@
+func NewKubernetesClientWithClient(client ListerClient) *KubernetesClient {
+	return &KubernetesClient{
+		client:           client,
+		close:            func() {},
+		closeM:           &sync.RWMutex{},
+		waitForCacheSync: func(context.Context) bool { return true },
+	}
+}
+
+// Close stops synchronization with the cluster. Subsequent calls to all of k's methods, excluding IsHealthy(), have
+// undefined behavior.
+func (k *KubernetesClient) Close() {
+	k.closeM.Lock()
+	defer k.closeM.Unlock()
+
+	// nil indicates Close() has been called for other methods.
+	k.close()
+	k.close = nil
+}
+
+// WaitForCacheSync waits for the internal client cache to synchronize.
+func (k *KubernetesClient) WaitForCacheSync(ctx context.Context) bool {
+	return k.waitForCacheSync(ctx)
+}
+
+// IsHealthy returns true until k.Close() is called. It is thread safe.
+func (k *KubernetesClient) IsHealthy(context.Context) bool {
+	k.closeM.RLock()
+	defer k.closeM.RUnlock()
+	return k.close != nil
+}
+
+// ByIP retrieves a hardware resource associated with ip.
+func (k *KubernetesClient) ByIP(ctx context.Context, ip string) (Hardware, error) {
+	var hw tinkv1alpha1.HardwareList
+	err := k.client.List(ctx, &hw, crclient.MatchingFields{
+		tink.HardwareIPAddrIndex: ip,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hw.Items) == 0 {
+		return nil, fmt.Errorf("no hardware with ip '%v'", ip)
+	}
+
+	if len(hw.Items) > 1 {
+		names := make([]string, len(hw.Items))
+		for i, item := range hw.Items {
+			names[i] = item.Name
+		}
+		return nil, fmt.Errorf("multiple hardware with ip '%v': [%v]", ip, strings.Join(names, ", "))
+	}
+
+	return NewK8sHardwareFromK8sObject(&hw.Items[0]), nil
+}
+
+// Watch is unimplemented.
+func (k *KubernetesClient) Watch(context.Context, string) (Watcher, error) {
+	return nil, errors.New("kubernetes client: watch is unimplemented")
+}
+
+// KuberneteSClientConfig used by the NewKubernetesClient function family.
+type KubernetesClientConfig struct {
+	*rest.Config
+	Namespace string
+}
+
+// NewKubernetesClientConfig loads the kubeconfig overriding it with kubeAPI.
+func NewKubernetesClientConfig(kubeconfig, kubeAPI string) (KubernetesClientConfig, error) {
+	kubeClientCfg := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{
+			ExplicitPath: kubeconfig,
+		},
+		&clientcmd.ConfigOverrides{
+			ClusterInfo: clientcmdapi.Cluster{
+				Server: kubeAPI,
+			},
+		},
+	)
+
+	config, err := kubeClientCfg.ClientConfig()
+	if err != nil {
+		return KubernetesClientConfig{}, err
+	}
+
+	namespace, _, err := kubeClientCfg.Namespace()
+	if err != nil {
+		return KubernetesClientConfig{}, err
+	}
+
+	return KubernetesClientConfig{
+		Config:    config,
+		Namespace: namespace,
+	}, nil
+}
+
+// NewK8sHardwareFromK8sObject creates an K8sHardware from tinkHardware.
+func NewK8sHardwareFromK8sObject(tinkHardware *tinkv1alpha1.Hardware) *K8sHardware {
```

Comment:
Tink's `pkg/convert` has conversions for templates and workflows. I think this is fine here, as we probably want to refactor the `Client` interface here to have some more generic hardware representation rather than Tink's eventually

Reply:
I noticed Boots is using similar types. I think we would benefit from some sort of back-end hardware module thats used among all services as we seem to have some dupe effort.